src/System.Management.Automation/utils/Encoding.cs

```diff
@@ -0,0 +1,399 @@
+        /// </summary>
+        public static FileEncoding GetFileEncodingFromFile(string path)
+        {
+            if (!File.Exists(path))
+            {
+                return FileEncoding.Default;
+            }
+
+            byte[] initialBytes = new byte[100];
+            int bytesRead = 0;
+
+            try
+            {
+                using (FileStream stream = System.IO.File.OpenRead(path))
+                {
+                    using (BinaryReader reader = new BinaryReader(stream))
+                    {
+                        bytesRead = reader.Read(initialBytes, 0, initialBytes.Length);
+                    }
+                }
+            }
+            catch (IOException)
+            {
+                return FileEncoding.Default;
+            }
+
+            // Test for four-byte preambles
+            string preamble = null;
+            FileEncoding foundEncoding = FileEncoding.Default;
+
+            if (bytesRead > 3)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1], initialBytes[2], initialBytes[3]);
+
+                if (encodingMap.TryGetValue(preamble, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Test for three-byte preambles
+            if (bytesRead > 2)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1], initialBytes[2]);
+                if (encodingMap.TryGetValue(preamble, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Test for two-byte preambles
+            if (bytesRead > 1)
+            {
+                preamble = String.Join("-", initialBytes[0], initialBytes[1]);
+                if (encodingMap.TryGetValue(preamble, out foundEncoding))
+                {
+                    return foundEncoding;
+                }
+            }
+
+            // Check for binary
+            string initialBytesAsAscii = System.Text.Encoding.ASCII.GetString(initialBytes, 0, bytesRead);
+            if (initialBytesAsAscii.IndexOfAny(nonPrintableCharacters) >= 0)
+            {
+                return FileEncoding.Byte;
+            }
+
+            // we couldn't determine anything from direct examination,
+            // return UTF8 without a BOM which should be good for both Windows and Non-Windows
+            return FileEncoding.UTF8NoBOM;
+        }
+
+        /// <summary>
+        /// Retrieve the PSDefaultFileEncoding preference value if set
+        /// <summary>
+        public static FileEncoding GetEncodingPreference(SessionState sessionState)
+        {
+            FileEncoding encodingPreference = FileEncoding.Unknown;
+            try
+            {
+                // It doesn't matter if this fails or throws, we will return unknown in that case
+                object tmp = sessionState.PSVariable.GetValue("PSDefaultFileEncoding");
```

Comment:
You should probably also test `sessionState != null` to avoid the expensive exception because I can see this method getting called with a null session state.